### PR TITLE
feat: initialize new package for signers

### PR DESCRIPTION
**Pull request by avasisht23**
# Pull Request Checklist

- [x] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [x] Did you update relevant docs? (docs are found in the `site` folder, see guidleines below)
- [x] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples? (e.g. `feat!: breaking change`)
- [x] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:fix`)
- [x] Is the base branch you're merging into `development` and not `main`?

### CODE REVIEW

**Comment by avasisht23** [comment]
Current dependencies on/for this PR:
* `development`
  * **PR #1** <a href="https://app.graphite.dev/github/pr/OMGWINNING/aa-sdk-stg-0/1?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/OMGWINNING/aa-sdk-stg-0/1?utm_source=stack-comment).

**Comment by moldy530** [comment]
any reason not to just do this directly in the aa-sdk? we could even have a branch called `aa-signers` if we want to stage changes further?

File: packages/signers/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  signTypedData: (params: SignTypedDataParameters) => Promise<Hex>;
+
+  authenticateUser: (params: AuthParams) => Promise<UserDetails>;
+
+  getUserDetails: () => Promise<UserDetails>;
+}
+
+export interface User<UserDetails> {
+  id: string;
+  details: UserDetails;
```
Comment (moldy530) [inline packages/signers/src/types.ts]: let's drop this

File: packages/signers/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import type { Hash, Hex, SignTypedDataParameters } from "viem";
+
+export interface SmartAccountSigner<Inner, AuthParams, UserDetails> {
```
Comment (moldy530) [inline packages/signers/src/types.ts]: I wonder if this should live in `aa-core` still. if this is here, then devs will ALWAYS need to install aa-signers which shouldn't be a requirement if devs want to bring their own in the end

File: packages/signers/src/types.ts (hunk 1 above)
Comment (avasisht23) [reply packages/signers/src/types.ts]: id + details, or just details?

File: packages/signers/src/types.ts (hunk 2 above)
Comment (avasisht23) [reply packages/signers/src/types.ts]: hmm good point. that makes sense.

File: packages/signers/src/types.ts (hunk 1 above)
Comment (moldy530) [reply packages/signers/src/types.ts]: just details, since we have a `getDetails` method

**Comment by avasisht23** [comment]
> any reason not to just do this directly in the aa-sdk? we could even have a branch called aa-signers if we want to stage changes further?

chatted live yesterday in sync: https://alchemyinsights.slack.com/archives/C04RHSDSP9U/p1699297061008119?thread_ts=1699074100.738799&cid=C04RHSDSP9U

tl;dr: this is a breaking change and requires 3rd party contributions. ideally this all gets in before merging to development. as we build out with turnkey/magic/lit, we don't want to hint to the other signers that we're partnering with them by having their impls in first.

that said, we could just merge in our own impls using the docs and then reach out to signers after to tweak. is that what you mean?

File: packages/signers/src/types.ts (hunk 2 above)
Comment (avasisht23) [reply packages/signers/src/types.ts]: this is actually more in line with how we did accounts

**Comment by moldy530** [comment]
Yea that's fine if we don't want to merge it into the `development` branch because it's a breaking change. Let's create a new branch called `aa-signer` in the main repo and use that as the base branch for all this dev. it'll be easier to track changes and get contributions

**Comment by denniswon** [comment]
Is the notion of User ”ALWAYS” relevant for signer? 

Then what about Local signers, Wallet Signers? 

I think we should align on the goal of having a separate aa-signer package.

In my opinion, it should be:
1. signer type for telemetry should not rely on arbitrary strings
2. there should be more dedicated process/place for 3rd party partners/signers to integrate with aa-sdk

**Comment by avasisht23** [comment]
> Is the notion of User ”ALWAYS” relevant for signer?

agreed, it's not. as I've been playing with integrating magic, I've realized this. let's discuss tomorrow.

right now I have something like this in aa-core/signer/types.ts to account for distinction between AuthSigner and Signer:

```
import type { Hash, Hex } from "viem";
import type { SignTypedDataParams } from "../account/types.js";

export interface AuthSmartAccountSigner<Inner, AuthParams, UserDetails>
  extends SmartAccountSigner {
  inner: Inner;
  user: SmartAccountUser<UserDetails> | undefined;

  authenticateUser: (params: AuthParams) => Promise<UserDetails>;

  getUserDetails: () => Promise<UserDetails>;
}

export interface SmartAccountSigner {
  signerType: string;

  getAddress: () => Promise<Hash>;

  signMessage: (msg: Uint8Array | Hex | string) => Promise<Hex>;

  signTypedData: (params: SignTypedDataParams) => Promise<Hex>;
}

export interface SmartAccountUser<UserDetails> {
  id: string | undefined;

  isAuthenticated: () => Promise<boolean>;

  getDetails: () => Promise<UserDetails>;
}

```
